File: 2024/day10.cpp
// https://adventofcode.com/2024/day/10

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>
#include <span>

// Highest value on the map is 9, so 11 can never be reached
constexpr let trailhead_char = '0';
constexpr let edge_char = '.';
constexpr let edge = u8{11};
constexpr let trail_end = u8{9};
using top_map_t = aoc::grid<u8>;

struct Input {
  top_map_t top_map;
  Vec<point> trailheads;
};

auto parse(String const& filename) -> Input {
  let char_map = aoc::read_char_grid(filename, {.padding = edge_char}).first;

  auto trailheads =
      aoc::views::indices_of(char_map) |
      stdv::filter(
          [&](usize i) { return char_map.at_index(i) == trailhead_char; }) |
      stdv::transform([&](usize i) { return char_map.position(i); }) |
      aoc::collect_vec<point>();

  auto heights = //
      char_map |
      stdv::transform([](char c) {
        return (c == edge_char) ? edge : aoc::to_number<u8>(c);
      }) |
      aoc::collect_vec<u8>();

  return {top_map_t{std::move(heights), char_map.num_rows(),
                    char_map.num_columns()},
          std::move(trailheads)};
}
fn test_map() -> top_map_t {
  return {std::vector{11, 11, 11, 11, 11, 11, //
                      11, 0,  1,  2,  3,  11, //
                      11, 1,  2,  3,  4,  11, //
                      11, 8,  7,  6,  5,  11, //
                      11, 9,  8,  7,  6,  11, //
                      11, 11, 11, 11, 11, 11} |
              aoc::collect_vec<u8>(),
          6, 6};
}

fn get_score(top_map_t const& top_map, point trailhead) -> usize {
  // Height strictly increases by 1 along every edge,
  // so the graph is a DAG and every path to a given cell has the same length -
  // reachability and "shortest distance" coincide,
  // so we can use Dijkstra to explore the whole trail from the trailhead
  let distances =
      aoc::shortest_distances_dijkstra(trailhead, [&](point current) {
        let current_height = top_map.at(current.y, current.x);
        return aoc::basic_neighbor_diffs |
               stdv::transform(
                   [current](point diff) { return current + diff; }) |
               stdv::filter([&top_map, current_height](point neighbor) {
                 return top_map.at(neighbor.y, neighbor.x) ==
                        current_height + 1;
               }) |
               aoc::dijkstra_uniform_neighbors_view();
      });

  return stdr::count_if(distances | stdv::keys, [&](point const& pos) {
    return top_map.at(pos.y, pos.x) == trail_end;
  });
}
static_assert(1 == get_score(test_map(), {1, 1}));

fn solve_case1(Input const& input) -> usize {
  return aoc::ranges::accumulate(
      input.trailheads | stdv::transform([&](point trailhead) {
        return get_score(input.top_map, trailhead);
      }),
      usize{});
}

struct SearchState {
  point pos;
  Vec<point> trail;

  constexpr bool operator==(SearchState const&) const = default;
};
template <>
struct std::hash<SearchState> {
  size_t operator()(SearchState const& state) const {
    auto combine = aoc::hash_combine{};
    combine(std::hash<point>{}(state.pos));
    for (point const& p : state.trail) {
      combine(std::hash<point>{}(p));
    }
    return combine.seed;
  }
};

fn get_rating(top_map_t const& top_map, point trailhead) -> usize {
  let start = SearchState{
      .pos = trailhead,
      .trail = {},
  };
  let num_trails = aoc::dfs_uniform(
      start,
      [&](SearchState const& state) {
        return top_map.at(state.pos.y, state.pos.x) == trail_end;
      },
      [&](SearchState const& state) {
        let current_height = top_map.at(state.pos.y, state.pos.x);
        return aoc::basic_neighbor_diffs |
               stdv::transform([&](point diff) { return state.pos + diff; }) |
               stdv::filter([&top_map, current_height](point neighbor) {
                 return top_map.at(neighbor.y, neighbor.x) ==
                        current_height + 1;
               }) |
               stdv::transform([&](point neighbor) {
                 auto new_state = state;
                 new_state.pos = neighbor;
                 new_state.trail.push_back(state.pos);
                 return new_state;
               });
      });
  return num_trails.at(start);
}

fn solve_case2(Input const& input) -> usize {
  return aoc::ranges::accumulate(
      input.trailheads | stdv::transform([&](point trailhead) {
        return get_rating(input.top_map, trailhead);
      }),
      usize{});
}

int main() {
  std::println("Part 1");
  let example = parse("day10.example");
  AOC_EXPECT_RESULT(1, solve_case1(example));
  let example2 = parse("day10.example2");
  AOC_EXPECT_RESULT(36, solve_case1(example2));
  let input = parse("day10.input");
  AOC_EXPECT_RESULT(652, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(81, solve_case2(example2));
  AOC_EXPECT_RESULT(1432, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2024/src/day10.rs
use aoc::dijkstra::DijkstraNeighborView;
use aoc::grid::{BASIC_NEIGHBOR_DIFFS, ConfigInput, Grid, Ipos};
use aoc::iter::CollectArrayVec;
use rustc_hash::FxHashMap;

const TRAILHEAD_CHAR: char = '0';
const EDGE_CHAR: char = '.';
// Highest value on the map is 9, so 11 can never be reached
const EDGE: u8 = 11;
const TRAIL_END: u8 = 9;

struct Input {
    top_map: Grid<u8>,
    trailheads: Vec<Ipos>,
}

fn parse(filename: &str) -> Input {
    let (char_map, _) = Grid::<char>::from_file_config(
        filename,
        ConfigInput {
            padding: Some(EDGE_CHAR),
            start_char: None,
            end_char: None,
        },
    );

    let trailheads: Vec<Ipos> = char_map
        .data
        .iter()
        .enumerate()
        .filter(|&(_, &c)| c == TRAILHEAD_CHAR)
        .map(|(i, _)| {
            let upos = char_map.position(i);
            Ipos::new(upos.x as isize, upos.y as isize)
        })
        .collect();

    let heights: Vec<u8> = char_map
        .data
        .iter()
        .map(|&c| {
            if c == EDGE_CHAR {
                EDGE
            } else {
                c.to_digit(10).unwrap() as u8
            }
        })
        .collect();

    let top_map = Grid::from_vec(heights, char_map.num_rows, char_map.num_columns);

    Input {
        top_map,
        trailheads,
    }
}

// Height strictly increases by 1 along every edge,
// so the graph is a DAG and every path to a given cell has the same length -
// reachability and "shortest distance" coincide,
// so we can use Dijkstra to explore the whole trail from the trailhead
fn get_score(top_map: &Grid<u8>, trailhead: Ipos) -> usize {
    let distances = aoc::dijkstra::shortest_distances(
        &trailhead,
        |_current: &Ipos| false,
        |&current: &Ipos| {
            let current_height = *top_map.get(current.y as usize, current.x as usize);
            BASIC_NEIGHBOR_DIFFS
                .iter()
                .map(move |&diff| current + diff)
                .filter(move |neighbor| {
                    *top_map.get(neighbor.y as usize, neighbor.x as usize) == current_height + 1
                })
                .dijkstra_uniform_neighbors()
        },
    );

    distances
        .keys()
        .filter(|&&pos| *top_map.get(pos.y as usize, pos.x as usize) == TRAIL_END)
        .count()
}

fn solve_case1(input: &Input) -> usize {
    input
        .trailheads
        .iter()
        .map(|&trailhead| get_score(&input.top_map, trailhead))
        .sum()
}

#[derive(Clone, PartialEq, Eq, Hash)]
struct SearchState {
    pos: Ipos,
    trail: Vec<Ipos>,
}

fn get_rating(top_map: &Grid<u8>, trailhead: Ipos) -> u64 {
    let start = SearchState {
        pos: trailhead,
        trail: Vec::new(),
    };
    let num_trails: FxHashMap<SearchState, u64> = aoc::algorithm::dfs_uniform(
        start.clone(),
        |state: &SearchState| *top_map.get(state.pos.y as usize, state.pos.x as usize) == TRAIL_END,
        |state: &SearchState| {
            let current_height = *top_map.get(state.pos.y as usize, state.pos.x as usize);
            BASIC_NEIGHBOR_DIFFS
                .iter()
                .map(|&diff| state.pos + diff)
                .filter(|&neighbor| {
                    *top_map.get(neighbor.y as usize, neighbor.x as usize) == current_height + 1
                })
                .map(|neighbor| {
                    let mut new_state = state.clone();
                    new_state.pos = neighbor;
                    new_state.trail.push(state.pos);
                    new_state
                })
                .collect_array_vec::<4>()
        },
    );
    num_trails[&start]
}

fn solve_case2(input: &Input) -> u64 {
    input
        .trailheads
        .iter()
        .map(|&trailhead| get_rating(&input.top_map, trailhead))
        .sum()
}

fn main() {
    println!("Part 1");
    let example = parse("day10.example");
    aoc::expect_result!(1, solve_case1(&example));
    let example2 = parse("day10.example2");
    aoc::expect_result!(36, solve_case1(&example2));
    let input = parse("day10.input");
    aoc::expect_result!(652, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(81, solve_case2(&example2));
    aoc::expect_result!(1432, solve_case2(&input));
}
